src/core/thread/mle.cpp

```diff
@@ -246,6 +252,7 @@ ThreadError Mle::Stop(void)
         mNetif.RemoveUnicastAddress(mLeaderAloc);
     }
 
+    mNetif.GetPendingDataset().StopDelayTimer();
```

Comment:
The scenario I'm trying to address is when a device is attached to the network, it should be in sync with the network.  In other words, if the device has a Pending Dataset that is not synchronized with the network, it should try to remain attached to that network so that it can provide the Pending Dataset to the Leader.  Starting the Delay Timer in `SetStateChild()` and `SetStateRouter()` can cause the device to apply its Pending Dataset before it's had a chance to synchronize it with the rest of the network.